src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -745,63 +746,204 @@ void StorageKafka2::dropReplica()
-                getNumber(keeper_to_use, *path_it / commit_file_name),
-                getNumber(keeper_to_use, *path_it / intent_file_name)};
+        TopicPartitions partitions_to_lock(ReplicaState.permanent_topics_assigned.begin(), ReplicaState.permanent_topics_assigned.begin() + can_lock_partitions);
+        for (const auto & partition_to_lock : partitions_to_lock)
+            createLocksInfo(keeper_to_use, permanent_locks, partition_to_lock);
 
-            LOG_TRACE(
-                log,
-                "Locked topic partition: {}:{} at offset {} with intent size {}",
-                tp_it->topic,
-                tp_it->partition_id,
-                lock_info.committed_offset.value_or(0),
-                lock_info.intent_size.value_or(0));
-            locks.emplace(TopicPartition(*tp_it), std::move(lock_info));
+        for (size_t i = can_lock_partitions; i < ReplicaState.permanent_topics_assigned.size(); ++i)
+            unlockTopicPartition(keeper_to_use, ReplicaState.permanent_topics_assigned[i]);
+    }
+    else
+    {
+        for (const auto & partition_to_lock : ReplicaState.permanent_topics_assigned)
+            createLocksInfo(keeper_to_use, permanent_locks, partition_to_lock);
+
+        for (; available_topics_it != available_topic_partitions.end(); ++available_topics_it)
+        {
+            if (permanent_locks.size() >= can_lock_partitions)
+                break;
+            createLocksInfo(keeper_to_use, permanent_locks, *available_topics_it, true);
         }
     }
 
+    ReplicaState::ReplicaStateData replica_state;
+    replica_state.permanent_topics_assigned.reserve(permanent_locks.size());
+    for (const auto & lock : permanent_locks)
+    {
+        replica_state.permanent_topics_assigned.push_back(
+            {
+                .topic = lock.first.topic,
+                .partition_id = lock.first.partition_id,
+                .offset = KafkaConsumer2::INVALID_OFFSET
+            }
+        );
+    }
+    replica_state.permanent_topic_partitions = replica_state.permanent_topics_assigned.size();
+
+    size_t can_lock_tmp_partitions = can_lock_partitions;
+    for (; available_topics_it != available_topic_partitions.end(); ++available_topics_it)
+    {
+        if (tmp_locks.size() >= can_lock_tmp_partitions)
+            break;
+        createLocksInfo(keeper_to_use, tmp_locks, *available_topics_it, true);
+    }
+    replica_state.tmp_topics_assigned.reserve(tmp_locks.size());
+    for (const auto & lock : tmp_locks)
+    {
+        replica_state.tmp_topics_assigned.push_back(
+            {
+                .topic = lock.first.topic,
+                .partition_id = lock.first.partition_id,
+                .offset = KafkaConsumer2::INVALID_OFFSET
+            }
+        );
+    }
+    replica_state.tmp_topic_partitions = replica_state.tmp_topics_assigned.size();
+    keeper_to_use.createOrUpdate(replica_path + "/topics_assigned", replica_state.toString(), zkutil::CreateMode::Persistent);
```

Comment:
I think this shouldn't be a persistent node. What if a replica is incorrectly removed, or clickhouse server stops and fails to restart? In that case these topic partitions won't be picked up by other replicas. Consider creating it as an emphemeral node or maybe you can completely remove it. You can always read `<replica path>/topics` and see which topic partitions are locked.